Keas.Mvc/Services/RolesSessionsManager.cs

```diff
@@ -0,0 +1,170 @@
+        /// <returns></returns>
+        public async Task<string[]> GetTeamRoleNames(string slug)
+        {
+            var roleContainer = GetRoleContainer();
+            roleContainer = await GetTeams(slug, roleContainer);
+            
+            return roleContainer.Teams.First(a => a.TeamName == slug).TeamRoles;
+        }
+
+        public async Task<string[]> GetTeamOrAdminRoleNames(string slug)
+        {
+            var roleContainer = GetRoleContainer();
+            roleContainer = await GetTeams(slug, roleContainer);
+            roleContainer = await GetSystemRoles(roleContainer);
+
+            var rolesNames = roleContainer.Teams.First(a => a.TeamName == slug).TeamRoles;
+
+
+            return rolesNames.Union(roleContainer.SystemRoles.TeamRoles).Distinct().ToArray();
+        }
+
+        public async Task<string[]> GetSystemRoleNames()
+        {
+            var roleContainer = GetRoleContainer();
+            roleContainer = await GetSystemRoles(roleContainer);
+
+            return roleContainer.SystemRoles.TeamRoles;
+        }
+
+        public void ClearSessionRoles()
+        {
+            _contextAccessor.HttpContext.Session.Remove(RolesSessionKey);
+        }
+
+        private RoleContainer GetRoleContainer()
+        {
+            var userId = _contextAccessor.HttpContext.User.Identity.Name;
+
+            var sessionResult = _contextAccessor.HttpContext.Session.GetString(RolesSessionKey);
+
+            var roleContainer = new RoleContainer();
+
+            if (sessionResult == null)
+            {
+                roleContainer = new RoleContainer();
+                roleContainer.UserId = userId;
+            }
+            else
+            {
+                try
+                {
+                    roleContainer = JsonConvert.DeserializeObject<RoleContainer>(sessionResult);
+                }
+                catch (Exception e)
+                {
+                    Log.Error(e.Message);
+                    roleContainer = new RoleContainer();
+                }
+
+                if (roleContainer.UserId != userId)
+                {
+                    _contextAccessor.HttpContext.Session.Remove(RolesSessionKey);
+                    roleContainer = new RoleContainer();
+                    roleContainer.UserId = userId;
+                }
+            }
+
+            return roleContainer;
+        }
+
+        private async Task<RoleContainer> GetSystemRoles(RoleContainer roleContainer)
+        {
+            var userId = _contextAccessor.HttpContext.User.Identity.Name;
+            var systemRole = roleContainer.SystemRoles;
+            if (systemRole == null)
+            {
+                systemRole = new Team
+                {
+                    TeamName = "System",
+                    TeamRoles = await _dbContext.SystemPermissions
+                        .AsNoTracking()
+                        .Where(a => a.UserId == userId).Select(a => a.Role.Name).Distinct().ToArrayAsync(),

```

Comment:
@jSylvestre It's safe to leave but I'm not sure you need asNoTracking() if you are going to .Select() project out.  There is nothing to track anyway

Reply:
Yeah, if the query is returning just a list of strings, he can probably leave it out. 

I'd even consider raw SQL for this part of the application.